pkg/util/connectgrpc/connectgrpc.go

```diff
@@ -115,17 +115,31 @@ func decodeRequest[Req any](req *httpgrpc.HTTPRequest) (*connect.Request[Req], e
 	return result, nil
 }
 
-func encodeRequest[Req any](req *connect.Request[Req]) (*httpgrpc.HTTPRequest, error) {
-	if req.Spec().Procedure == "" {
-		return nil, errors.New("cannot encode a request with empty procedure")
+type connectURLCtxKey struct{}
+
+func WithProcedure(ctx context.Context, u string) context.Context {
+	return context.WithValue(ctx, connectURLCtxKey{}, u)
+}
+
+func ProcedureFromContext(ctx context.Context) string {
+	s, _ := ctx.Value(connectURLCtxKey{}).(string)
+	return s
+}
+
+func encodeRequest[Req any](ctx context.Context, req *connect.Request[Req]) (*httpgrpc.HTTPRequest, error) {
+	url := req.Spec().Procedure
+	if url == "" {
+		if url = ProcedureFromContext(ctx); url == "" {
```

Comment:
Unsure which one of the two should take precedence, might be useful to override in a test using the context, but it is really arbitrary.